modules/encoders/x86/bmp_polyglot.rb

```diff
@@ -0,0 +1,423 @@
+  ; get bit loop is done                                                            |
+  pop ecx               ; restore the byte counter                                  |
+  mov byte [edi], al    ; move the newly decoded byte to its final destination      |
+  inc edi               ; increment the destination pointer                         |
+                        ;                                                           |
+  dec ecx               ; decrement the byte counter                                |
+  jne get_byte          ; ----------------------------------------------------------/
+
+  ; get byte loop is done
+  popad                 ; restore all registers
+
+=end
+
+# calculate the smallest increase of a 32-bit little endian integer which is
+# also a valid x86 jmp opcode of the specified minimum size.
+class SizeCalculator
+  def initialize(size, minimum_jump)
+    @original_size = size
+    raise if minimum_jump < 0 || minimum_jump > 0xff
+    @minimum_jump = minimum_jump
+  end
+
+  def calculate
+    possibles = []
+    size = new_size_long
+    possibles << size unless size.nil?
+    size = new_size_short
+    possibles << size unless size.nil?
+    return if possibles.length == 0
+    possibles.min
+  end
+
+  def new_size_long
+    size = [ @original_size ].pack('V').unpack('CCCC')
+
+    0.upto(2) do |i|
+      b0 = size[i]
+      b1 = size[i + 1]
+      b2 = size[i + 2].to_i
+      b3 = size[i + 3].to_i
+      b4 = size[i + 4].to_i
+      min_jmp = (@minimum_jump - 5 - i)
+
+      if [b2, b3, b4].inject{ |sum, x| sum + x } > 0  # this jmp would be too large
+        if b0 > 0xfd
+          size = increment_size(size, i)
+        end
+        size[i] = round_up_to_nop(b0)
+        next
+      end
+
+      if b0 > 0xe9
+        if b0 > 0xfd
+          size = increment_size(size, i)
+        end
+        size[i] = round_up_to_nop(b0)
+      else
+        size[i] = 0xe9
+        b1 = min_jmp if b1 < min_jmp
+        size[i + 1] = b1
+        return size.pack('CCCC').unpack('V')[0]
+      end
+    end
+  end
+
+  def new_size_short
+    return if @minimum_jump > 0x81  # short won't make it in this case (0x7f + 0.upto(2).to_a.max)
+    size = [ @original_size ].pack('V').unpack('CCCC')
+
+    0.upto(2) do |i|
+      b0 = size[i]
+      b1 = size[i + 1]
+      min_jmp = (@minimum_jump - 2 - i)
+
+      if b0 > 0xeb
+        if b0 > 0xfd
+          size = increment_size(size, i)
+        end
+        size[i] = round_up_to_nop(b0)
+      else
+        size[i] = 0xeb
+        if b1 > 0x7f
+          b1 = min_jmp
+          size = increment_size(size, i + 1)
+        elsif b1 < min_jmp
+          b1 = min_jmp
+        end
+        size[i + 1] = b1
+        return size.pack('CCCC').unpack('V')[0]
+      end
+    end
+  end
+
+  def size_to_jmp(size)
+    jmp = 0
+    packed = [ size, 0 ].pack('VV')
+
+    until [ "\xe9", "\xeb" ].include?(packed[0])
```

Comment:
No because it needs to be a jmp (either short or long), a call could be used but would modify the stack. The size does not need to start with e9 or eb though, it could be something like `90 eb 60 00` where the first byte is a nop.